File: ios/VCInvestmentDiscovery/ViewModels/CompanyProfileViewModel.swift
import Foundation
import Combine

@MainActor
class CompanyProfileViewModel: ObservableObject {
    @Published var company: Company?
    @Published var financialData: FinancialData?
    @Published var socialData: SocialData?
    @Published var marketData: MarketData?
    @Published var relatedTrends: [Trend] = []
    
    private let dataService: DataService
    private var cancellables = Set<AnyCancellable>()
    
    init(dataService: DataService, companyId: UUID) {
        self.dataService = dataService
        setupBindings()
        loadCompanyProfile(companyId: companyId)
    }
    
    private func setupBindings() {
        dataService.$company
            .assign(to: \.company, on: self)
            .store(in: &cancellables)
        
        dataService.$financialData
            .assign(to: \.financialData, on: self)
            .store(in: &cancellables)
        
        dataService.$socialData
            .assign(to: \.socialData, on: self)
            .store(in: &cancellables)
        
        dataService.$marketData
            .assign(to: \.marketData, on: self)
            .store(in: &cancellables)
        
        dataService.$relatedTrends
            .assign(to: \.relatedTrends, on: self)
            .store(in: &cancellables)
    }
    
    func loadCompanyProfile(companyId: UUID) {
        Task {
            await dataService.fetchCompany(companyId)
            await dataService.fetchFinancialData(companyId)
            await dataService.fetchSocialData(companyId)
            await dataService.fetchMarketData(companyId)
            await dataService.fetchRelatedTrends(companyId)
        }
    }
    
    func refreshCompanyProfile() {
        guard let companyId = company?.id else { return }
        loadCompanyProfile(companyId: companyId)
    }
    
    func updateFinancialData(_ newData: FinancialData) {
        guard let companyId = company?.id else { return }
        Task {
            await dataService.updateFinancialData(newData, for: companyId)
            self.financialData = newData
        }
    }
    
    func updateSocialData(_ newData: SocialData) {
        guard let companyId = company?.id else { return }
        Task {
            await dataService.updateSocialData(newData, for: companyId)
            self.socialData = newData
        }
    }
    
    func updateMarketData(_ newData: MarketData) {
        guard let companyId = company?.id else { return }
        Task {
            await dataService.updateMarketData(newData, for: companyId)
            self.marketData = newData
        }
    }
    
    func calculateInvestmentPotential() -> Double {
        // TODO: Implement a sophisticated algorithm for calculating investment potential
        // This is a placeholder implementation
        let financialScore = financialData?.revenue ?? 0
        let socialScore = Double(socialData?.twitterFollowers ?? 0)
        let marketScore = marketData?.marketSize ?? 0
        
        let weightedScore = (financialScore * 0.4) + (socialScore * 0.3) + (marketScore * 0.3)
        return min(max(weightedScore / 1_000_000, 0), 100) // Normalize to 0-100 scale
    }
}

File: android/app/src/main/java/com/vcinvestmentdiscovery/viewmodels/CompanyProfileViewModel.kt
package com.vcinvestmentdiscovery.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import com.vcinvestmentdiscovery.data.repository.CompanyRepository
import com.vcinvestmentdiscovery.data.models.*
import com.vcinvestmentdiscovery.util.Result
import java.util.UUID

@HiltViewModel
class CompanyProfileViewModel @Inject constructor(
    private val repository: CompanyRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(CompanyProfileUiState())
    val uiState: StateFlow<CompanyProfileUiState> = _uiState.asStateFlow()

    fun loadCompanyProfile(companyId: UUID) {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch {
            try {
                val company = repository.getCompany(companyId)
                val financialData = repository.getFinancialData(companyId)
                val socialData = repository.getSocialData(companyId)
                val marketData = repository.getMarketData(companyId)
                val relatedTrends = repository.getRelatedTrends(companyId)

                _uiState.value = CompanyProfileUiState(
                    company = company,
                    financialData = financialData,
                    socialData = socialData,
                    marketData = marketData,
                    relatedTrends = relatedTrends,
                    isLoading = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "An error occurred while loading the company profile"
                )
            }
        }
    }

    fun refreshCompanyProfile() {
        uiState.value.company?.id?.let { companyId ->
            loadCompanyProfile(companyId)
        }
    }

    fun updateFinancialData(newData: FinancialData) {
        viewModelScope.launch {
            try {
                val result = repository.updateFinancialData(newData)
                if (result is Result.Success) {
                    _uiState.value = _uiState.value.copy(financialData = newData)
                } else {
                    _uiState.value = _uiState.value.copy(
                        error = (result as Result.Error).message
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "An error occurred while updating financial data"
                )
            }
        }
    }

    fun updateSocialData(newData: SocialData) {
        viewModelScope.launch {
            try {
                val result = repository.updateSocialData(newData)
                if (result is Result.Success) {
                    _uiState.value = _uiState.value.copy(socialData = newData)
                } else {
                    _uiState.value = _uiState.value.copy(
                        error = (result as Result.Error).message
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "An error occurred while updating social data"
                )
            }
        }
    }

    fun updateMarketData(newData: MarketData) {
        viewModelScope.launch {
            try {
                val result = repository.updateMarketData(newData)
                if (result is Result.Success) {
                    _uiState.value = _uiState.value.copy(marketData = newData)
                } else {
                    _uiState.value = _uiState.value.copy(
                        error = (result as Result.Error).message
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "An error occurred while updating market data"
                )
            }
        }
    }

    fun calculateInvestmentPotential(): Double {
        val financialData = _uiState.value.financialData
        val socialData = _uiState.value.socialData
        val marketData = _uiState.value.marketData

        // TODO: Implement a sophisticated algorithm for calculating investment potential
        // This is a placeholder implementation and should be refined with domain experts
        val financialScore = financialData?.let { (it.revenue + it.profit) / it.valuation } ?: 0.0
        val socialScore = socialData?.let { (it.followers + it.engagement) / 1000.0 } ?: 0.0
        val marketScore = marketData?.let { it.marketShare * it.growthRate } ?: 0.0

        return (financialScore * 0.5 + socialScore * 0.3 + marketScore * 0.2) * 100
    }
}